ENH: implement stringdtype <-> timedelta roundtrip casts

ngoldbaum opened the pull request:
I missed this cast before adding stringdtype to numpy. It turns out the pandas test suite expects this round-trip cast to work.

Note that this differs from the existing casts between timedelta and unicode string, which are (IMO) useless and kind of broken:

```
In [1]: import numpy as np

In [2]: arr = np.array([123, 456], dtype='m8[s]')

In [3]: arr.astype(str)
Out[3]: array(['123 seconds', '456 seconds'], dtype='<U21')

In [4]: arr = np.array([1234567891011121314, 456], dtype='m8[s]')

In [5]: arr.astype(str)
Out[5]: array(['1234567891011121314 s', '456 seconds'], dtype='<U21')
```

Not totally sure where `"U21"` is coming from, but even if there wasn't any truncation this still wouldn't roundtrip:

```
In [6]: arr = np.array([123, 456], dtype='m8[s]')

In [7]: arr.astype(str).astype('m8[s]')
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[7], line 1
----> 1 arr.astype(str).astype('m8[s]')

ValueError: Could not convert object to NumPy timedelta
```

This happens because the numpy code for parsing strings as datetimes [only supports parsing strings as integers directly](https://github.com/numpy/numpy/blob/a53b0b3777a966554add1313e22b625638282192/numpy/_core/src/multiarray/datetime.c#L2535) (side note, the use of `strtol` there is almost certainly buggy under non-C locales, it also doesn't do any overflow checking).

I've chosen here to not touch those code paths and instead always output the string representation of integers or "NaT" and leave it up to the caller to deal with the timedelta unit metadata. This has the nice property that timedeltas roundtrip automatically as long as you specify the unit:

```
In [1]: import numpy as np

In [2]: arr = np.array([1234567891011121314, 456], dtype='m8[s]')

In [3]: arr.astype(np.dtypes.StringDType())
Out[3]: array(['1234567891011121314', '456'], dtype=StringDType())

In [4]: arr.astype(np.dtypes.StringDType()).astype(arr.dtype)
Out[4]: array([1234567891011121314,                 456], dtype='timedelta64[s]')
```

ngoldbaum committed:
ENH: fix parsing of empty and 'NaT' strings in string to datetime cast

ngoldbaum committed:
ENH: implement casts between timedelta and stringdtype

ngoldbaum commented:
I added the 2.0 release milestone to this one because it would be nice to have in before branching.

Maybe @mhvk or @seberg could have a look if you have time?

No worries if it doesn't make it in before the RC, I can just xfail the pandas test that needs this.

mhvk commented on numpy/_core/src/multiarray/stringdtype/casts.c:
```diff
@@ -1394,6 +1406,143 @@ static PyType_Slot dt2s_slots[] = {
+                    "Failed to load string in string to datetime cast");
+            goto fail;
+        }
+        if (is_null || is_nat_string(&s)) {
```
It does not seem logical that if the string is not missing itself, but equal to `NaT`, that then the outcome depends on the `StringDType` - I think it should always be `NPY_DATETIME_NAT`. Maybe easiest to just separate out the behaviour for `is_null` and `is_nat_string`, deal with `is_null` first and then check `is_nat_string` (so that if the default string is `NaT`, that get's caught appropriately)

mhvk commented on numpy/_core/src/multiarray/stringdtype/casts.c:
```diff
@@ -1394,6 +1406,143 @@ static PyType_Slot dt2s_slots[] = {
+        }
+
+        PyObject *pystr = PyUnicode_FromStringAndSize(s.buf, s.size);
+
```
Total nitpick, but might as well remove some of these empty lines...

mhvk commented on numpy/_core/src/multiarray/stringdtype/casts.c:
```diff
@@ -1394,6 +1406,143 @@ static PyType_Slot dt2s_slots[] = {
+            }
+        }
+
+        if (int_to_stringbuf((long long)*in, out, allocator) != 0) {
```
this should be an `else if`, no? Otherwise the `NaT` gets interpreted as a integer here.

mhvk commented on numpy/_core/tests/test_stringdtype.py:
```diff
@@ -809,18 +810,38 @@ def test_ufunc_multiply(dtype, string_list, other, other_dtype, use_out):
+    a = np.array(input_data, dtype=input_dtype)
 
     has_na = hasattr(dtype, "na_object")
     is_str = isinstance(getattr(dtype, "na_object", None), str)
```
I think it would be good to test NaT also with a default string of `NaT` (could be done in the next test too).

More generally, `NaT` should round-trip except when the default string is used and is not equal to "NaT"

mhvk reviewed (commented):
Mostly looks good, but I think two logic errors - and a request to update the tests so that at least the first one would be found.

ngoldbaum committed:
BUG: fix issues pointed out in review, add tests

ngoldbaum committed:
ENH: add support for null strings in assert_array_compare

ngoldbaum commented:
I implemented your suggestions, thanks for pointing out those issues! Adding the tests you suggested led me find that `assert_array_equal` didn't know how to deal with nan-like null strings, so I added support for that. Happy to add more extensive tests for that if you think we need it.

mhvk commented on numpy/_core/src/multiarray/stringdtype/casts.c:
```diff
@@ -1394,6 +1422,152 @@ static PyType_Slot dt2s_slots[] = {
+
+static int
+string_to_timedelta(PyArrayMethod_Context *context, char *const data[],
+                                       npy_intp const dimensions[], npy_intp const strides[],
```
Indentation?

mhvk commented on numpy/_core/src/multiarray/stringdtype/casts.c:
```diff
@@ -1394,6 +1422,152 @@ static PyType_Slot dt2s_slots[] = {
+            else if (NpyString_pack_null(allocator, out_pss) < 0) {
+                npy_gil_error(
+                        PyExc_MemoryError,
+                        "Failed to pack string in datetime to string cast");
```
datetime -> timedelta

mhvk commented on numpy/_core/tests/test_stringdtype.py:
```diff
@@ -842,7 +863,41 @@ def test_datetime_cast(dtype):
+        dt_array = arr.astype('M8[s]')
+        td_array = arr.astype('m8[s]')
+        assert_array_equal(
+            dt_array, np.array([NaT_dt] * arr.size, dtype='M8[s]'))
```
In principle, `assert_array_equal` can compare to a scalar, so `assert_array_equal(dt_array, NaT_dt)` should "just work".

mhvk commented on numpy/_core/tests/test_stringdtype.py:
```diff
@@ -842,7 +863,41 @@ def test_datetime_cast(dtype):
+        else:
+            output_object = na_object
+
+        for arr in [dt_array, td_array]:
```
One could parametrize over `datetime` and `timedelta`, but, really...

mhvk reviewed (approved):
Some really small bits left. Since they are trivial, will approve now, but might as well correct the copy & paste error in the error message at least.

ngoldbaum committed:
MNT: respond to review comments

mhvk commented:
Super, looks all OK!

ngoldbaum commented:
Doctest failures are because the scipy nightly wheel is broken at the moment.

Thanks for looking this one over @mhvk!
